### PR TITLE
Add GSoC 2022 page

### PR DESCRIPTION
**Pull request by XVilka**

### CODE REVIEW

File: content/gsoc/2022.md [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+- Write the test cases for Rizin regression tests and improve the results.
+- Implement necessary commands and APIs in Rizin for visual representation of the IL tree
+- Implement standard and graph views in Cutter for the IL output
+
+### Skills
+Participant should know C and bits of C++ as well as be familiar with basics of the program analysis. Having an
+experience with other intermediate language, SAT/SMT, and mathematical logic is a plus.
+
+### Difficulty
+Medium
```
Comment (wargio) [inline content/gsoc/2022.md:97]: Hard/Very Hard

File: content/gsoc/2022.md [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+is to finish the migration of the existing architectures that still use ESIL or add a RZIL support
+for the architectures that hadn't any uplifting at all.
+
+### Tasks
+
+- Improvement a RZIL uplfiting for any non-trivial architecture, preferably that is supported by ESIL already
+- Improve the integration with analysis (variables and types differences) for the chosen architecture
+- Write the test cases for Rizin regression tests and improve the results.
+- Implement necessary commands and APIs in Rizin for visual representation of the IL tree
+- Implement standard and graph views in Cutter for the IL output
```
Comment (wargio) [inline content/gsoc/2022.md:90]: maybe say to use the rz-tracetest from @thestr4ng3r

File: content/gsoc/2022.md [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+
+### Difficulty
+Hard
+
+### Benefits for the participant
+Participant will understand the hurdles of multithreaded programming, data syncronization, locks and debugging of such code.
+
+### Assess requirements for midterm/final evaluation
+- 1st term: `RzTypes` module exists and contains all relevant code.
+- Final term: Tree-sitter based C parser is implemented and integrated into the analysis framework.
```
Comment (wargio) [inline content/gsoc/2022.md:192]: i think this is wrong. is this a copy of last year one?

File: content/gsoc/2022.md [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+
+## Rewriting GPL-only code (Medium project)
+
+Currently some of the Rizin main features rely on the GPL-only code copied from binutils or GCC.
+The goal is to rewrite all this code from GPL-only to LGPL or any other less restrictive license.
+It is quite important for better adoption of Rizin as a library in other FOSS and commercial
+projects.
+
+## Tasks
+1. Rewrite C++ demangler to no use GPL-only code
```
Comment (wargio) [inline content/gsoc/2022.md:210]:
```suggestion
1. Rewrite C++ demangler to and remove the GPL code
```

Comment (wargio) [reply content/gsoc/2022.md:210]: maybe implement the GNUv2 C++ demangler if too easy?

File: content/gsoc/2022.md [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+	- Hexadecimal view
+	- Disassembly view
+	- Graph view
+	- Pseudocode view
+
+### Skills
+The participant should be comfortable with the C++ language, and be familiar with Qt framework
+
+### Difficulty
+Advanced
```
Comment (wargio) [inline content/gsoc/2022.md:468]: i think this is medium. hex diffing is slightly more difficult than asm/grap/pseudo diffing.
i would be nice to have the results of rz-diff for function similarities on 2 bins into cutter

**Review by wargio — CHANGES_REQUESTED** [review]

File: content/gsoc/2022.md [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+7. Parallelize some of the RzAnalysis function using the threading API
+
+### Skills
+Participant should know C as well as have the experience of developing multithreaded applications.
+
+### Difficulty
+Hard
+
+### Benefits for the participant
+Participant will understand the hurdles of multithreaded programming, data syncronization, locks and debugging of such code.
```
Comment (officialcjunior) [inline content/gsoc/2022.md:188]: A small typo here:

```suggestion
Participant will understand the hurdles of multithreaded programming, data synchronization, locks and debugging of such code.
```

File: content/gsoc/2022.md (hunk 4 above)
Comment (XVilka) [reply content/gsoc/2022.md:210]: Yes

File: content/gsoc/2022.md [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+- [SWIG Website](http://www.swig.org/)
+- [SWIG 4.0 Documentation](http://www.swig.org/Doc4.0/SWIGDocumentation.html)
+- [Small PoC of bindings generated in Rizin's build system](https://github.com/rizinorg/rizin/pull/285)
+- [Article about Rizin's build system design](https://rizin.re/posts/why-meson/)
+
+# Cutter
+
+## Plugins and Python High Level API (Medium project)
+
+We currently don't have API almost for plugin authors to use. We need to improve a lot of things about our Plugins support and take it few steps ahead.
```
Comment (thestr4ng3r) [inline content/gsoc/2022.md:341]:
```suggestion
Our current public API to be used by plugin authors is somewhat limited. We need to improve a lot of things about our Plugins support and take it few steps ahead.
```

File: content/gsoc/2022.md [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+This task is only about improving the Python interface in Cutter, specifically its graphical user interface components. For a task about exposing Rizin's API for disassembly, analysis and other purposes, see the Rizin bindings task above.
```
Comment (thestr4ng3r) [inline content/gsoc/2022.md:342]:
```suggestion
This task is only about improving the public C++ and Python interface of Cutter, specifically its graphical user interface components. For a task about exposing Rizin's API for disassembly, analysis and other purposes, see the Rizin bindings task above.
```

File: content/gsoc/2022.md [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+- thestr4ng3r
+- Megabeets
+
+### Links/Resources
+- [SDB Module/API for Cutter Python/Jupyter integration](https://github.com/rizinorg/cutter/issues/1778)
+- [Jupyter plugin for Cutter](https://github.com/rizinorg/cutter-jupyter)
+
+## Multi-Tasking and Event-driven architecture (Extended deadline)
+
+Cutter is a reverse engineering framework that is powered by Rizin. The information it gets about functions, strings, imports, and the analysis are all performed in Rizin and displayed in Cutter. Currently, Cutter is pulling information from Rizin only on demand. This is problematic because sometimes the user performs changes (via plugins, the console widget, and more) that are affecting the information from Rizin, but Cutter doesn't know about these changes to apply the to the UI. For example, if  a user will define a new function in a Python script or via the console widget by using the Rizin command `af @ <addr>`, Cutter will now show this new function in the Functions widget until the user will refresh the interface manually (edit -> Refresh Contents).
```
Comment (thestr4ng3r) [inline content/gsoc/2022.md:377]:
```suggestion
The information Cutter gets about functions, strings, imports, and the analysis are all performed in Rizin and only displayed in Cutter. Currently, it is pulling most information from Rizin only on demand. This is problematic because sometimes the user performs changes (via plugins, the console widget, and more) that are affecting the information from Rizin, but Cutter doesn't know about these changes to apply the to the UI. For example, if a user will define a new function in a Python script or via the console widget by using the Rizin command `af @ <addr>`, Cutter will not show this new function in the Functions widget until the user will refresh the interface manually (Edit -> Refresh Contents).
The goal of this task is to use an event-driven architecture to overcome this limitation.
```

File: content/gsoc/2022.md [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+
+### Tasks
+The overall implementation of this task should start from Rizin by adding events to many of the functions. This can be done using `rz_events`. For example, add an even for function creating, for section creation, for flag deletion, for name changed, and more
+
+- Add events to all the relevant functions inside Rizin
+- Add support for these events in Cutter and refresh and update the relevant widgets per each event
+- Support analysis in the background and allow the user to start its session while Rizin is analyzing (see [#1856](https://github.com/rizinorg/cutter/issues/1856), [#1574](https://github.com/rizinorg/cutter/issues/1574))
+
+### Skills
+The participant should be comfortable with the C++ for Cutter and C for Rizin. The participant should be familiar with Qt framework.
```
Comment (thestr4ng3r) [inline content/gsoc/2022.md:389]:
```suggestion
The participant should be comfortable with the C++ for Cutter and C for Rizin. They should also be familiar with Qt framework. Experience in GUI code architecture, for example using functional reactive programming or Elm-like approaches is a plus.
```

File: content/gsoc/2022.md [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+It will allow to work on big files effortlessly in Cutter, will improve analysis quality as well.
+
+### Assess requirements for midterm/final evaluation
+- 1st term: Implement events everywhere in the relevant places across Rizin code and event-driven
+	interaction with Cutter.
+- Final term: Add support for the Cutter interface refresh based on the events from Rizin, implement analysis in background.
+
+### Mentors
+- thestr4ng3r
+- Karliss
```
Comment (thestr4ng3r) [inline content/gsoc/2022.md:407]: Karliss isn't on the mentor list (I think he is not available this year, is he?)

File: content/gsoc/2022.md [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+
+## Heap viewer (Medium project)
```
Comment (thestr4ng3r) [inline content/gsoc/2022.md:409]: Wasn't that partially finished last year?

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]

File: content/gsoc/2022.md (hunk 1 above)
Comment (wargio) [reply content/gsoc/2022.md:97]: the hard part is to verify.

File: content/gsoc/2022.md (hunk 12 above)
Comment (yossizap) [reply content/gsoc/2022.md:409]: There are major performance issues with the windows implementation and [Cutter integration wasn't completed](https://github.com/rizinorg/cutter/pull/2723). This could only be a medium project if support for other allocators is included.

File: content/gsoc/2022.md [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+# Introduction
+
+This year is a second we participate as a fork - Rizin, effectively continuing the tradition since the 2015 year (as a radare2 project).
+
+## Mentors
+
+Members of the Rizin and Cutter core teams have volunteered to guide participants for GSoC’22. They were already guiding the participants for the GSoC and RSoC in past years. Please feel free to reach out to any of them in case you need any help in selecting a project.
+
+- **Anton Kochkov** Mattermost: xvilka -- [@akochkov](https://twitter.com/akochkov)
+- **Riccardo Shirone** Mattermost: ret2libc
```
Comment (ret2libc) [inline content/gsoc/2022.md:21]:
```suggestion
- **Riccardo Schirone** Mattermost: ret2libc
```

File: content/gsoc/2022.md [hunk 14]
```diff
@@ -0,0 +1,600 @@
...
+- [ESIL to RZIL conversion tracking issue](https://github.com/rizinorg/rizin/issues/2080)
+- [Cutter: IL output and graph visual representation](https://github.com/rizinorg/rizin/issues/1133)
+
+## Debug information handling improvements (Medium project)
+
+Rizin already supports most of the DWARF and PDB features, including cross-platform parsing of both.
+But there are huge chunks are missing in terms of the integration an source line information.
+Current task would be to improve parsing support of both, add compatibility with `debuginfod`,
+and enhance breakpoint integration, and variable/structure printing in the debugging mode
+with the source line information.
```
Comment (ret2libc) [inline content/gsoc/2022.md:129]:
```suggestion
Rizin already supports most of the DWARF and PDB features, including cross-platform parsing of both.
However information are usually just printed to aid the reverse engineering process, but they are not 
actually used at their best. For example, you can't use them to configure a breakpoint, nor they can be
used to access variables within a function during debugging. Moreover, it is becoming more and more 
common to store DWARF information in separate files, either shipped as separate file or downloaded 
on the fly with `debuginfod`. Rizin does not support these kind of DWARF files yet.

Your task would be to improve the parsing support of both by fixing smaller bugs, add support for
separate DWARF files and `debuginfod` and enhance breakpoint integration and variable/structure
printing in debugging mode with the source information gathered from DWARF/PDB.
```

File: content/gsoc/2022.md [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+
+### Task
+1. Support loading DWARF information from separate files and `debuginfod`
+2. Unify source lines/types information access for DWARF, PDB, dSYM
```
Comment (ret2libc) [inline content/gsoc/2022.md:133]:
```suggestion
2. Unify source lines/types information access for DWARF, PDB, dSYM and refactor/fix parsing code as necessary
```

File: content/gsoc/2022.md [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+3. Integrate source line and types/variables information with the analysis
+4. Integrate source line and types/variables with printing with `p` commands in the debug mode
+5. Integrate source line and types/variables with breakpoint commands and APIs
+6. Parsing performance improvements
+
+### Skills
+Participant should know C as well as have the experience of debugging with GDB or LLDB.
```
Comment (ret2libc) [inline content/gsoc/2022.md:140]:
```suggestion
* Good knowledge of the C language
* Some experience in debugging with GDB or LLDB
* Basic knowledge of at least one of the following formats: ELF, DWARF, PDB, PE
```

File: content/gsoc/2022.md [hunk 17]
```diff
@@ -0,0 +1,600 @@
...
+
+### Difficulty
+Hard
+
+### Benefits for the participant
+Participant will understand how high-level features of debuggers work as well as gain skills in the field of software architecture of a large, modular C project.
+
+### Assess requirements for midterm/final evaluation
+- 1st term: `debuginfod` and source line information refactoring are implemented
+- Final term: Integration of variable information with the analysis and debug printing is implemented
```
Comment (ret2libc) [inline content/gsoc/2022.md:150]:
```suggestion
- Final term: Integration of variable information with the debug and printing commands is implemented
```

File: content/gsoc/2022.md [hunk 18]
```diff
@@ -0,0 +1,600 @@
...
+### Links/Resources
+- [Loading debug information from `debuginfod`](https://github.com/rizinorg/rizin/issues/1285)
+- [Unify code of source information access for DWARF, PDB, dSYM](https://github.com/rizinorg/rizin/issues/907)
+- [Debuginfod](https://sourceware.org/elfutils/Debuginfod)
+- [Debian Debuginfod](https://wiki.debian.org/Debuginfod)
+- [Fedora Debuginfod](https://fedoraproject.org/wiki/Debuginfod)
+- [DWARF-labeled issues](https://github.com/rizinorg/rizin/labels/DWARF)
+- [PDB-labaled issues](https://github.com/rizinorg/rizin/labels/PDB)
+
+## Thread-safety and multithreading (Medium project)
```
Comment (ret2libc) [inline content/gsoc/2022.md:166]: Not a big fan of this project... I think we may have better things to propose :(

File: content/gsoc/2022.md [hunk 19]
```diff
@@ -0,0 +1,600 @@
...
+
+## Heap viewer (Medium project)
+
+We already have a nice heap (and memory map) parser and visualizer in Rizin (`dm` and `dmh` commands). After debugging becomes a first-class citizen in cutterland it would be awesome to have memory map and heap visualizations.
+
+### Task
+ - Expose Rizin API/commands for Cutter to use for visualization
+ - Design and implement heap navigation and inspection widgets
+ - Provide the integration with current debugging mode in Cutter
+ - Make the implementation work with both local (native) and remote debugging modes
```
[the remark this answers is not in the document]
Comment (XVilka) [reply content/gsoc/2022.md:418]: Yes, good idea

File: content/gsoc/2022.md (hunk 18 above)
Comment (XVilka) [reply content/gsoc/2022.md:173]: Please share your ideas then.

**Comment by XVilka** [comment]
@Maijin good ideas, except `binwalk -Me`.

**TLDR;** better not to do this.

I think, using the binwalk itself is preferred since Rizin doesn't aim to work with filesystems and thousands of the different formats. We removed `m` (mount) commands after all, and it's better to focus on the code reverse engineering features rather than implementing inferior copy of the binwalk. I am aware of many original binwalk problems, but Rizin is not the place to implement a better version of it. If anything, I would have implemented hypothetical **binwalk-ng** in Rust language, reusing a lot of useful existing libraries, like [lzma-rs](https://github.com/gendx/lzma-rs/).

File: content/gsoc/2022.md (hunk 1 above)
Comment (wargio) [reply content/gsoc/2022.md:106]: let's keep it like this and judge based on the choosen arch by the student.

File: content/gsoc/2022.md [hunk 20]
```diff
@@ -0,0 +1,618 @@
...
+This year is the second time we participate as a fork - Rizin, effectively continuing the tradition since the year 2015 (as the radare2 project).
+
+## Mentors
+
+Members of the Rizin and Cutter core teams have volunteered to guide participants for GSoC’22. They were already guiding the participants for the GSoC and RSoC in past years. Please feel free to reach out to any of them in case you need any help in selecting a project.
+
+- **Anton Kochkov** Mattermost: xvilka -- [@akochkov](https://twitter.com/akochkov)
+- **Riccardo Schirone** Mattermost: ret2libc
+- **Florian Märkl** Mattermost/Telegram: @thestr4ng3r -- [@thestr4ng3r](https://twitter.com/thestr4ng3r)
+- **Antide Petit** IRC/Telegram: xarkes -- [@xarkes\_](https://twitter.com/xarkes_)
```
Comment (xarkes) [inline content/gsoc/2022.md:23]: Due to my current involvement in the project, I don't think I should mentor this year, so I think you can remove this line :)

Comment (wargio) [inline content/gsoc/2022.md:23]:
```suggestion
```
